File: jad_deob/sources/p000/BitmapFont.java
package p000;

import java.io.DataInputStream;
import java.io.IOException;
import javax.microedition.lcdui.Graphics;
import javax.microedition.lcdui.Image;

/* renamed from: a */
public final class BitmapFont {

    /* renamed from: a */
    private static int[] f0a = null;

    /* renamed from: e */
    private static byte[] f1e = null;

    /* renamed from: f */
    private static byte[] f2f = null;

    /* renamed from: a */
    private byte fontHeight;

    /* renamed from: a */
    private int bmpWidth;

    /* renamed from: a */
    private C0004e f5a;

    /* renamed from: a */
    private byte[] f6a;

    /* renamed from: a */
    private C0004e[] f7a;

    /* renamed from: b */
    private byte f8b = 1;

    /* renamed from: b */
    private int bmpHeight;

    /* renamed from: b */
    private byte[] f10b;

    /* renamed from: c */
    private int f11c;

    /* renamed from: c */
    private byte[] f12c;

    /* renamed from: d */
    private int f13d;

    /* renamed from: d */
    private byte[] f14d;

    public BitmapFont(boolean z) {
        String str;
        if (f1e == null) {
            DataInputStream dataInputStream = new DataInputStream(DrawingUtils.loadResourceByName("/t"));
            f1e = new byte[225];
            f2f = new byte[225];
            short s = 0;
            for (int i = 0; i < 225; i++) {
                try {
                    byte[] bArr = f1e;
                    s = (short) (s + dataInputStream.readShort());
                    bArr[i] = (byte) s;
                } catch (IOException e) {
                    return;
                }
            }
            short s2 = 0;
            for (int i2 = 0; i2 < 225; i2++) {
                byte[] bArr2 = f2f;
                s2 = (short) (s2 + dataInputStream.readShort());
                bArr2[i2] = (byte) s2;
            }
        }
        if (z) {
            str = "/sans80b.png";
            this.f6a = f2f;
            this.f8b = 0;
            this.f11c = 4;
        } else {
            str = "/sans80r.png";
            this.f6a = f1e;
            this.f8b = 1;
            this.f11c = 7;
        }
        Image a = DrawingUtils.loadImageResource(str);
        if (a != null) {
            this.fontHeight = 11;
            this.bmpWidth = a.getWidth();
            this.bmpHeight = a.getHeight();
            this.f7a = new C0004e[this.f11c];
            for (int i3 = 0; i3 < this.f11c; i3++) {
                this.f7a[i3] = new C0004e(null);
            }
            this.f13d = 0;
            int[] iArr = new int[(this.bmpWidth * this.bmpHeight)];
            a.getRGB(iArr, 0, this.bmpWidth, 0, 0, this.bmpWidth, this.bmpHeight);
            this.f14d = new byte[(this.bmpWidth * this.bmpHeight)];
            int length = iArr.length;
            for (int i4 = 0; i4 < length; i4++) {
                this.f14d[i4] = (byte) (7 - ((iArr[i4] & 255) >> 5));
            }
            this.f10b = new byte[225];
            this.f12c = new byte[225];
            byte b = 0;
            byte b2 = 0;
            for (int i5 = 0; i5 < 225; i5++) {
                if (this.f6a[i5] + b2 >= this.bmpWidth) {
                    b = (byte) (b + 1);
                    b2 = 0;
                }
                this.f10b[i5] = b2;
                this.f12c[i5] = b;
                b2 = (byte) (b2 + this.f6a[i5]);
            }
            byte[] bArr3 = this.f6a;
            bArr3[0] = (byte) (bArr3[0] - this.f8b);
        }
    }

    /* renamed from: a */
    private static char convertCharToFontIndex(char c) {
        if (c >= ' ' && c <= 255) {
            return (char) (c - ' ');
        }
        switch (c) {
            case 8211:
                return 'v';
            case 8212:
                return 'w';
            case 8216:
                return 'q';
            case 8217:
                return 'r';
            case 8220:
                return 's';
            case 8221:
                return 't';
            case 8226:
                return 'u';
            case 8230:
                return 'e';
            default:
                return 224;
        }
    }

    /* renamed from: a */
    private int m1a(String str, int i, int i2) {
        int i3 = 0;
        int i4 = i + i2;
        if (i4 > str.length()) {
            i4 = str.length();
        }
        while (i < i4) {
            i3 += this.f6a[convertCharToFontIndex(str.charAt(i))] + this.f8b;
            i++;
        }
        return i3;
    }

    /* renamed from: a */
    private int m2a(Graphics graphics, char c, int i, int i2) {
        char a = convertCharToFontIndex(c);
        if (a != 0) {
            if (this.f5a.f78a != null) {
                graphics.drawRegion(m3a(this.f12c[a]), this.f10b[a], 0, this.f6a[a], this.fontHeight, 0, i, i2, 0);
            } else {
                graphics.drawRGB(m5a(this.f12c[a]), this.f10b[a], this.bmpWidth, i, i2, this.f6a[a], this.fontHeight, false);
            }
        }
        return this.f6a[a] + this.f8b;
    }

    /* renamed from: a */
    private Image m3a(byte b) {
        if (this.f5a.f78a[b] == null) {
            int i = this.bmpWidth * this.fontHeight;
            int i2 = b * i;
            for (int i3 = 0; i3 < i; i3++) {
                f0a[i3] = this.f5a.f76a[this.f14d[i3 + i2]];
            }
            this.f5a.f78a[b] = Image.createRGBImage(f0a, this.bmpWidth, this.fontHeight, false);
        }
        return this.f5a.f78a[b];
    }

    /* renamed from: a */
    private void m4a(int i, int i2) {
        if (!(this.f5a != null && this.f5a.f75a == i && this.f5a.f80b == i2)) {
            for (int i3 = 0; i3 < this.f11c; i3++) {
                if (i == this.f7a[i3].f75a && i2 == this.f7a[i3].f80b) {
                    this.f5a = this.f7a[i3];
                    return;
                }
            }
            this.f13d = (this.f13d + 1) % this.f11c;
            this.f5a = this.f7a[this.f13d];
            this.f5a.f78a = new Image[10];
            if (f0a == null || f0a.length < this.bmpWidth * this.fontHeight) {
                f0a = new int[(this.bmpWidth * this.fontHeight)];
            }
            this.f5a.f75a = i;
            this.f5a.f80b = i2;
            this.f5a.f76a = m6a(i, i2);
        }
    }

    /* renamed from: a */
    private int[] m5a(byte b) {
        if (!this.f5a.f79a[b]) {
            this.f5a.f79a[b] = true;
            int i = this.fontHeight * this.bmpWidth;
            int i2 = b * i;
            if (this.f5a.f77a[b] == null) {
                this.f5a.f77a[b] = new int[(this.bmpWidth * this.fontHeight)];
            }
            int[] iArr = (int[]) this.f5a.f77a[b];
            int[] iArr2 = this.f5a.f76a;
            for (int i3 = 0; i3 < i; i3++) {
                iArr[i3] = iArr2[this.f14d[i3 + i2]];
            }
        }
        return (int[]) this.f5a.f77a[b];
    }

    /* renamed from: a */
    private static int[] m6a(int i, int i2) {
        int[] iArr = new int[8];
        int i3 = (i >> 16) & 255;
        int i4 = (i >> 8) & 255;
        int i5 = (i >> 0) & 255;
        int i6 = (i2 >> 16) & 255;
        int i7 = (i2 >> 8) & 255;
        int i8 = (i2 >> 0) & 255;
        iArr[0] = i2;
        for (int i9 = 1; i9 < 7; i9++) {
            iArr[i9] = (((((8 - i9) * i6) + (i3 * i9)) >> 3) << 16) | (((((8 - i9) * i7) + (i4 * i9)) >> 3) << 8) | (((((8 - i9) * i8) + (i5 * i9)) >> 3) << 0);
        }
        iArr[7] = i;
        return iArr;
    }

    /* renamed from: a */
    public final int strFontLengthThing(String str) {
        return m1a(str, 0, str.length());
    }

    /* renamed from: a */
    public final int mo2a(char[] cArr, int i, int i2) {
        int i3 = 0;
        int i4 = i + i2;
        while (i < i4) {
            i3 += this.f6a[convertCharToFontIndex(cArr[i])] + this.f8b;
            i++;
        }
        return i3;
    }

    /* renamed from: a */
    public final short mo3a() {
        return (short) this.fontHeight;
    }

    /* renamed from: a */
    public final void mo4a(Graphics graphics, String str, int i, int i2, int i3, int i4) {
        mo5a(graphics, str, i, i2, i3, i4, Integer.MAX_VALUE, i4);
    }

    /* renamed from: a */
    public final void mo5a(Graphics graphics, String str, int i, int i2, int i3, int i4, int i5, int i6) {
        int length = str.length();
        if (length != 0) {
            int a = (i3 & 1) != 0 ? i - (strFontLengthThing(str) / 2) : (i3 & 8) != 0 ? i - strFontLengthThing(str) : i;
            if ((i3 & 32) != 0) {
                i2 -= this.fontHeight;
            }
            int clipX = graphics.getClipX();
            int clipY = graphics.getClipY();
            int clipWidth = graphics.getClipWidth();
            int clipHeight = graphics.getClipHeight();
            m4a(graphics.getColor(), i4);
            int i7 = 0;
            int i8 = i5;
            while (i7 < length) {
                int a2 = m2a(graphics, str.charAt(i7), a, i2);
                if (a + a2 > i8) {
                    m4a(graphics.getColor(), i6);
                    int max = Math.max(i8, clipX);
                    graphics.setClip(max, clipY, (clipWidth + clipX) - max, clipHeight);
                    i8 = Integer.MAX_VALUE;
                    i7--;
                } else {
                    a += a2;
                }
                i7++;
            }
            graphics.setClip(clipX, clipY, clipWidth, clipHeight);
        }
    }

    /* renamed from: a */
    public final void mo6a(Graphics graphics, char[] cArr, int i, int i2, int i3, int i4, int i5) {
        int i6 = i + i2;
        m4a(graphics.getColor(), i5);
        while (i < i6) {
            i3 += m2a(graphics, cArr[i], i3, i4);
            i++;
        }
    }
}

File: OBML/js/BitmapFont.js
class C0004e {
  /** @type {number} */
  f75a;

  /** @type {int[]} */
  f76a = [];

  /** @type {Object[]}  */
  f77a = [];

  /** @type {Image[]}  */
  f78a = [];

  /** @type {boolean[]}  */
  f79a = [];

  /** @type {int}  */
  f80b;

  /**
   *
   * @param {C0019t} tVar
   */
  constructor(tVar) {}
}

class BitmapFont {
  static f1e = null;
  static f2f = null;

  /** @type {C0004e} */
  f5a = new C0004e(); // TODO !

  fontHeight;
  bmpWidth;
  bmpHeight;

  static async loadFontsCache() {
    const response = await fetch("assets/font_map");
    const raw_data = await response.arrayBuffer();
    const data = new Uint16Array(raw_data);
    BitmapFont.f1e = new Uint8Array(225);
    BitmapFont.f2f = new Uint8Array(225);
    let s = 0;
    for (let i = 0; i < 225; i++) {
      s = s + data[i];
      BitmapFont.f1e[i] = s & 0xff;
    }
    let s2 = 0;
    for (let i = 0; i < 225; i++) {
      s2 = s2 + data[i + 225];
      BitmapFont.f2f[i] = s2 & 0xff;
    }
  }

  async loadFontImage(name) {
    return new Promise((resolve, reject) => {
      let img = new Image();
      img.onload = () => resolve(img);
      img.onerror = reject;
      img.src = name;
    });
  }

  pushFontImage(a) {
    this.fontHeight = 11;
    this.bmpWidth = a.width;
    this.bmpHeight = a.height;

    this.f7a = new Array(this.f11c);

    for (let i3 = 0; i3 < this.f11c; i3++) {
      this.f7a[i3] = new C0004e(null);
    }

    this.f13d = 0;
    const iArr = new Array(this.bmpWidth * this.bmpHeight);
    const canvas = document.createElement("canvas");
    canvas.width = this.bmpWidth;
    canvas.height = this.bmpHeight;
    const context = canvas.getContext("2d");
    context.drawImage(a, 0, 0);
    // context
    //   .getImageData(0, 0, this.bmpWidth, this.bmpHeight)
    //   .data.forEach((value, i4) => {
    //     iArr[i4] = value;
    //   });

    const imageData = context.getImageData(0, 0, this.bmpWidth, this.bmpHeight);
    const pixels = [];
    for (let i = 0; i < imageData.data.length; i += 4) {
      const r = imageData.data[i],
        g = imageData.data[i + 1],
        b = imageData.data[i + 2],
        a = imageData.data[i + 3];

      // TODO: unsure about the order, doc say argb
      iArr[i / 4] = (a << 24) + (r << 16) + (g << 8) + b;
    }

    this.f14d = new Uint8Array(this.bmpWidth * this.bmpHeight);
    for (let i4 = 0; i4 < iArr.length; i4++) {
      // Look like it's stripping the alpha but touching the blue ?
      this.f14d[i4] = (7 - ((iArr[i4] & 255) >> 5)) & 0xff;
    }
    this.f10b = new Uint8Array(225);
    this.f12c = new Uint8Array(225);
    let b = 0,
      b2 = 0;

    for (let i5 = 0; i5 < 225; i5++) {
      if (this.f6a[i5] + b2 >= this.bmpWidth) {
        b = (b + 1) & 0xff;
        b2 = 0;
      }
      this.f10b[i5] = b2 & 0xff;
      this.f12c[i5] = b & 0xff;
      b2 = (b2 + this.f6a[i5]) & 0xff;
    }
    this.f6a[0] = (this.f6a[0] - this.f8b) & 0xff;
  }

  async createFonts(z) {
    let str;

    if (z) {
      str = "assets/sans80b.png";
      this.f6a = BitmapFont.f2f;
      this.f8b = 0;
      this.f11c = 4;
    } else {
      str = "assets/sans80r.png";
      this.f6a = BitmapFont.f1e;
      this.f8b = 1;
      this.f11c = 7;
    }

    const a = await this.loadFontImage(str);
    this.pushFontImage(a);
  }

  constructor(z) {}

  static convertCharToFontIndex(c) {
    if (c.charCodeAt(0) >= " ".charCodeAt(0) && c.charCodeAt(0) <= 255) {
      return String.fromCharCode(c.charCodeAt(0) - " ".charCodeAt(0));
    }
    switch (c.charCodeAt(0)) {
      case 8211:
        return "v";
      case 8212:
        return "w";
      case 8216:
        return "q";
      case 8217:
        return "r";
      case 8220:
        return "s";
      case 8221:
        return "t";
      case 8226:
        return "u";
      case 8230:
        return "e";
      default:
        return String.fromCharCode(224);
    }
  }

  async init(z) {
    if (!BitmapFont.f1e) {
      // Load font data from binary file
      await BitmapFont.loadFontsCache();
      await this.createFonts(z);
    } else {
      await this.createFonts(z);
    }
    return this;
  }

  mo1a(str) {
    return this.m1a(str, 0, str.length);
  }

  // Get String length ?
  m1a(str, start_index, str_len) {
    let size = 0;
    let max_size = start_index + str_len;
    if (max_size > str.length) {
      max_size = str.length;
    }
    while (start_index < max_size) {
      size +=
        this.f6a[
          BitmapFont.convertCharToFontIndex(str.charAt(start_index)).charCodeAt(
            0
          )
        ] + this.f8b;
      start_index++;
    }
    return size;
  }

  m2a(graphics, c, i, i2) {
    let a = BitmapFont.convertCharToFontIndex(c).charCodeAt(0);
    if (a !== 0) {
      if (!!this.f5a?.f78a?.length) {
        graphics.drawRegion(
          this.m3a(this.f12c[a]),
          this.f10b[a],
          0,
          this.f6a[a],
          this.fontHeight,
          0,
          i,
          i2,
          0
        );
      } else {
        graphics.drawRGB(
          this.m5a(this.f12c[a]),
          this.f10b[a],
          this.bmpWidth,
          i,
          i2,
          this.f6a[a],
          this.fontHeight,
          false
        );
      }
    }
    return this.f6a[a] + this.f8b;
  }

  m3a(b) {
    if (this.f5a.f78a[b] == null) {
      const i = this.bmpWidth * this.fontHeight;
      const i2 = b * i;
      const f0a = new Array(i);

      for (let i3 = 0; i3 < i; i3++) {
        f0a[i3] = this.f5a.f76a[this.f14d[i3 + i2]];
      }
      this.f5a.f78a[b] = new Image();
      this.f5a.f78a[b].src = "data:image/png;base64," + btoa(f0a.join(","));
    }

    return this.f5a.f78a[b];
  }

  m5a(b) {
    if (this.f5a?.f79a && this.f5a?.f79a[b]) {
      return this.f5a.f77a[b];
    }
    this.f5a.f79a[b] = true;
    var i = this.fontHeight * this.bmpWidth;
    var i2 = b * i;
    if (this.f5a.f77a[b] == null) {
      this.f5a.f77a[b] = new Array(this.bmpWidth * this.fontHeight);
    }
    var iArr = this.f5a.f77a[b];
    var iArr2 = this.f5a.f76a;
    for (var i3 = 0; i3 < i; i3++) {
      iArr[i3] = iArr2[this.f14d[i3 + i2]];
    }
  }
}
